optuna/study/study.py

```diff
@@ -928,6 +936,18 @@ def _pop_waiting_trial_id(self) -> Optional[int]:
 
         return None
 
+    def _should_skip_enqueue(self, params: Dict[str, Any]) -> bool:
+
+        for trial in self.get_trials(deepcopy=False):
+            trial_params = trial.system_attrs.get("fixed_params", trial.params)
+            if trial_params == params:
```

Comment:
That was my initial approach, but implementation turns out a bit complicated, since we need to check for types now, so I figured precision in cpython `dict_equal` is enough for this API. But I see where `json` encoding-decoding can introduce float representation related bugs, so if you don't mind slightly complicated  type checking, I'm fine with bringing back implementation from 56126a21e78957f6834193cbb4be490368fa2ba0 (with less bugs this time :^)).

Reply:
Sounds good :+1: I think slightly complicated type checking is acceptable for this purpose.

Furthermore, `float('nan') == float('nan')` will be false actually. It caused a bug that is reported at #3556. To be honest, I'm not sure we should support the edge case. However, given that we've fixed #3556, it seems that we need to support the `float('nan')` input here.

Reply:
Done in cf6810e645ca24a21cd060017f1fc329fa5f38d5 (somehow awkwardly). PTAL @c-bata.